doc/developer/design/20230330_recursion_limit.md

```diff
@@ -0,0 +1,187 @@
+
+Note that having separate limits for separate WMR clauses in one query is especially important for the **compositionality** of views. Let's say that Alice writes a view that has just a single WMR. Then Bob writes another query that uses Alice's view, and has a WMR itself. The view is inlined, so Bob's query now has two WMRs. But Bob shouldn't need to care about the internals of Alice's view, and might not even know that the view has a WMR, so he thinks his WMR is the only one. And then he sets a limit for the entire query, which will unexpectedly mess up Alice's WMR that is inlined from inside the view. This problem is somewhat relevant for all the three use cases, but especially for 3.
+
+Queries with multiple WMR clauses can be classified into 3 categories:
+- *Independent* WMR clauses (whose results eventually flow together into, e.g., a union or join),
+- *Sequenced* WMR clauses, i.e., where one WMR clause uses the results of another,
+- *Nested* WMR clauses.
+
+For nested WMR clauses, we have the complication that the limit of the inner WMR might need to depend on state from the outer WMR; for example, on the current iteration number of the outer query. This can happen easily in both the 2. and 3. use cases. For example, one might want to debug an inner WMR in a specific iteration of the outer WMR. We can extend C. to accommodate for a dependency on arbitrary outer state if we allow not just a constant expression as the limit. Or, if we just want to extend C. specifically to accommodate a dependency on the outer iteration number (instead of arbitrary outer state), then we might allow a pointstamp to be the limit (this would be my preference for a first implementation). D. is more flexible even without special extensions: the WHERE clause in the inner WMR can naturally refer to arbitrary outer state. If we specifically want to access the outer iteration number, then we can save that iteration number in a new binding of the outer WMR.
+
+### Queries with multiple WMR clauses -- implementation difficulties
+
+C. and D. are, unfortunately, non-trivial to implement and would also introduce a maintenance burden. The problem is that in naive implementations of these features, limit settings would not survive optimizer transforms, e.g., `NormalizeLets` or `PredicatePushdown`.
+
+#### C.
+
+For C., the main difficulty is that `NormalizeLets` merges *sequenced* or *independent* WMR clauses (and sometimes *nested* ones as well, but only when the inner one is not really recursive, I think) into one WMR clause. Setting just one limit of this merged WMR clause would be wrong. One solution would be to have a per-let-binding limit in MIR and LIR (and in rendering), so that different let bindings that came from different WMR clauses with differing limits could retain their differing limits. This needs some improvements in `NormalizeLets`:
```

Comment:
> only when the inner one is not really recursive

If this is correct we can probably just propagate the limit from the other block (or maybe `limit + 1`, I'm not sure).

Reply:
I'm planning to set it to infinite.